src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -91,8 +91,15 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
                               callbackInfo->FileSize.QuadPart);
     };
 
+    auto vfs = reinterpret_cast<OCC::VfsCfApi *>(callbackInfo->CallbackContext);
+    Q_ASSERT(vfs->metaObject()->className() == QByteArrayLiteral("OCC::VfsCfApi"));
+
+    const auto requestId = QString::number(callbackInfo->TransferKey.QuadPart, 16);
     const auto sendTransferInfo = [=](QByteArray &data, qint64 offset) {
         qDebug(lcCfApiWrapper) << "Send transfer info. Offset:" << offset;
+        if (vfs->hydrationCanceled(requestId)) {
+            return;
```

Comment:
why are you not calling `doneWithHydration` here ?
I miss the explanation as to why the clean up will still be done in that case

Reply:
`doneWithHydration` needs to be called after the cfapi is done with hydrating. Otherwise, it is no possible to check if a job was cancelled or not. The hydration jobs hold the information if a job was cancelled or not. In the past, these jobs freed themselves after they were done. It could happen that a hydration job is done BEFORE the cfapi had processed all data. In that case, it is not possible to see if the hydration job was cancelled or not, because the hydration job was already gone. Therefore the hydration job needs to stay alive until the cfapi is finally done. `doneWithHydration` needs to be called regardless if the job was cancelled or not, to allow the vfs to free the hydration job. `sendTransferInfo` will be called multiple times. I'm not sure I understand how `doneWithHydration` could be called in there. Does that answer your question?